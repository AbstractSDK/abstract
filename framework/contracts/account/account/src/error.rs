use cosmwasm_std::StdError;

use manager::error::ManagerError;
use proxy::error::ProxyError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AccountError {
    #[error("{0}")]
    Std(#[from] StdError),

    // #[error("{0}")]
    // Abstract(#[from] AbstractError),

    // #[error("{0}")]
    // AbstractSdk(#[from] AbstractSdkError),

    // #[error("{0}")]
    // Admin(#[from] AdminError),
    #[error(transparent)]
    Manager(#[from] ManagerError),

    #[error(transparent)]
    Proxy(#[from] ProxyError),
}
