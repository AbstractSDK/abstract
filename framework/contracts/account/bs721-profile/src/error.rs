use cosmwasm_std::StdError;
use cw_controllers::AdminError;
use cw_ownable::OwnershipError;
use cw_utils::PaymentError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ContractError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    OwnershipError(#[from] OwnershipError),
    #[error("{0}")]
    Payment(#[from] PaymentError),

    #[error("{0}")]
    Admin(#[from] AdminError),

    #[error("{0}")]
    Base(#[from] bs721_base::ContractError),

    #[error("NameNotFound")]
    NameNotFound {},

    #[error("AddressAlreadyMapped")]
    AddressAlreadyMapped {},

    #[error("RecordNameAlreadyExists")]
    RecordNameAlreadyExists {},

    #[error("RecordNameEmpty")]
    RecordNameEmpty {},

    #[error("RecordNameTooLong")]
    RecordNameTooLong {},

    #[error("RecordValueTooLong")]
    RecordValueTooLong {},

    #[error("RecordValueEmpty")]
    RecordValueEmpty {},

    #[error("UnauthorizedVerification")]
    UnauthorizedVerification {},

    #[error("Invalid Metadata")]
    InvalidMetadata {},

    #[error("Unauthorized: Not contract creator or admin")]
    UnauthorizedCreatorOrAdmin {},

    #[error("TooManyRecords max: {max}")]
    TooManyRecords { max: u32 },

    #[error("NotImplemented")]
    NotImplemented {},
}
