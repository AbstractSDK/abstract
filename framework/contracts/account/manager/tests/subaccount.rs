use abstract_core::{
    manager::SubAccountIdsResponse,
    objects::{gov_type::GovernanceDetails, AccountId},
    PROXY,
};
use abstract_integration_tests::{create_default_account, AResult};
use abstract_interface::*;
use abstract_manager::error::ManagerError;
use abstract_testing::OWNER;
use cosmwasm_std::{to_json_binary, wasm_execute, Addr};
use cw_orch::{contract::Deploy, prelude::*};

#[test]
fn creating_on_subaccount_should_succeed() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;
    let sub_accounts = account.manager.sub_account_ids(None, None)?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            // only one sub-account and it should be account_id 2
            sub_accounts: vec![2]
        }
    );
    take_storage_snapshot!(chain, "creating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn updating_on_subaccount_should_succeed() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;

    // Subaccount should have id 2 in this test, we try to update the config of this module
    let account_contracts = get_account_contracts(&deployment.version_control, AccountId::local(2));
    let new_desc = "new desc";
    account_contracts
        .0
        .update_info(Some(new_desc.to_string()), None, None)?;

    assert_eq!(
        Some(new_desc.to_string()),
        account_contracts.0.info()?.info.description
    );
    take_storage_snapshot!(chain, "updating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn proxy_updating_on_subaccount_should_succeed() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    let proxy_address = account.proxy.address()?;
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;

    // Subaccount should have id 2 in this test, we try to update the config of this module
    let (sub_manager, _) = get_account_contracts(&deployment.version_control, AccountId::local(2));
    let new_desc = "new desc";

    // We call as the proxy, it should also be possible
    sub_manager
        .call_as(&proxy_address)
        .update_info(Some(new_desc.to_owned()), None, None)?;

    assert_eq!(
        Some(new_desc.to_string()),
        sub_manager.info()?.info.description
    );

    take_storage_snapshot!(chain, "proxy_updating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn recursive_updating_on_subaccount_should_succeed() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;

    // Subaccount should have id 2 in this test, we try to update the config of this module
    let account_contracts = get_account_contracts(&deployment.version_control, AccountId::local(2));

    // We call as the manager, it should also be possible
    account_contracts.0.create_sub_account(
        vec![],
        "My subsubaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;
    let account_contracts = get_account_contracts(&deployment.version_control, AccountId::local(3));
    let new_desc = "new desc";

    account_contracts
        .0
        .call_as(&sender)
        .update_info(Some(new_desc.to_string()), None, None)?;

    assert_eq!(
        Some(new_desc.to_string()),
        account_contracts.0.info()?.info.description
    );

    take_storage_snapshot!(chain, "recursive_updating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn installed_app_updating_on_subaccount_should_succeed() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;
    let first_proxy_addr = account.proxy.address()?;

    let mock_app = Addr::unchecked("mock_app");
    account
        .proxy
        .call_as(&account.manager.address()?)
        .add_modules(vec![mock_app.to_string()])?;

    let (sub_manager, _sub_proxy) =
        get_account_contracts(&deployment.version_control, AccountId::local(2));
    let new_desc = "new desc";

    // recover address on first proxy
    account.proxy.set_address(&first_proxy_addr);
    // adding mock_app to whitelist on proxy

    // We call as installed app of the owner-proxy, it should also be possible
    account
        .proxy
        .call_as(&mock_app)
        .module_action(vec![wasm_execute(
            sub_manager.addr_str()?,
            &abstract_core::manager::ExecuteMsg::UpdateInfo {
                name: None,
                description: Some(new_desc.to_owned()),
                link: None,
            },
            vec![],
        )?
        .into()])?;

    assert_eq!(
        Some(new_desc.to_string()),
        sub_manager.info()?.info.description
    );
    take_storage_snapshot!(chain, "installed_app_updating_on_subaccount_should_succeed");

    Ok(())
}

#[test]
fn sub_account_move_ownership() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let new_owner = Addr::unchecked("new_owner");
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    // Store manager address, it will be used for querying
    let manager_addr = account.manager.address()?;

    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;
    let sub_accounts = account.manager.sub_account_ids(None, None)?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            // only one sub-account and it should be account_id 2
            sub_accounts: vec![2]
        }
    );

    let sub_account = AbstractAccount::new(&deployment, AccountId::local(2));
    sub_account
        .manager
        .propose_owner(GovernanceDetails::Monarchy {
            monarch: new_owner.to_string(),
        })?;

    // Make sure it's not updated until claimed
    let sub_accounts: SubAccountIdsResponse = chain.query(
        &abstract_core::manager::QueryMsg::SubAccountIds {
            start_after: None,
            limit: None,
        },
        &manager_addr,
    )?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            sub_accounts: vec![2]
        }
    );

    // Claim ownership
    sub_account.manager.call_as(&new_owner).execute(
        &abstract_core::manager::ExecuteMsg::UpdateOwnership(cw_ownable::Action::AcceptOwnership),
        None,
    )?;
    let account = AbstractAccount::new(&deployment, AccountId::local(1));

    // After claim it's updated
    let sub_accounts = account.manager.sub_account_ids(None, None)?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            sub_accounts: vec![]
        }
    );
    take_storage_snapshot!(chain, "sub_account_move_ownership");

    Ok(())
}

#[test]
fn account_move_ownership_to_sub_account() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    Abstract::deploy_on(chain.clone(), sender.to_string())?;
    abstract_integration_tests::manager::account_move_ownership_to_sub_account(chain)?;
    Ok(())
}

#[test]
fn sub_account_move_ownership_to_sub_account() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;

    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;
    let sub_account = AbstractAccount::new(&deployment, AccountId::local(2));
    let sub_manager_addr = sub_account.manager.address()?;
    let sub_proxy_addr = sub_account.proxy.address()?;

    let new_account = create_default_account(&deployment.account_factory)?;
    new_account.manager.create_sub_account(
        vec![],
        "My second subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;

    // sub-accounts state updated
    let sub_ids = new_account.manager.sub_account_ids(None, None)?;
    assert_eq!(sub_ids.sub_accounts, vec![4]);

    let new_account_sub_account = AbstractAccount::new(&deployment, AccountId::local(4));
    let new_governance = GovernanceDetails::SubAccount {
        manager: sub_manager_addr.to_string(),
        proxy: sub_proxy_addr.to_string(),
    };
    new_account_sub_account
        .manager
        .propose_owner(new_governance.clone())?;
    let new_account_sub_account_manager = new_account_sub_account.manager.address()?;

    let sub_account = AbstractAccount::new(&deployment, AccountId::local(2));
    let mock_module = Addr::unchecked("mock_module");
    sub_account
        .proxy
        .call_as(&sub_manager_addr)
        .add_modules(vec![mock_module.to_string()])?;
    sub_account
        .proxy
        .call_as(&mock_module)
        .module_action(vec![wasm_execute(
            new_account_sub_account_manager,
            &abstract_core::manager::ExecuteMsg::UpdateOwnership(
                cw_ownable::Action::AcceptOwnership,
            ),
            vec![],
        )?
        .into()])?;

    // sub-accounts state updated
    let sub_ids = sub_account.manager.sub_account_ids(None, None)?;
    assert_eq!(sub_ids.sub_accounts, vec![4]);
    let new_account = AbstractAccount::new(&deployment, AccountId::local(3));
    // removed from the previous owner as well
    let sub_ids = new_account.manager.sub_account_ids(None, None)?;
    assert_eq!(sub_ids.sub_accounts, Vec::<u32>::new());

    let new_account_sub_account = AbstractAccount::new(&deployment, AccountId::local(4));
    let info = new_account_sub_account.manager.info()?.info;
    assert_eq!(new_governance, info.governance_details.into());
    take_storage_snapshot!(chain, "sub_account_move_ownership_to_sub_account");

    Ok(())
}

#[test]
fn account_move_ownership_to_falsy_sub_account() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    let proxy_addr = account.proxy.address()?;

    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;
    let sub_account = AbstractAccount::new(&deployment, AccountId::local(2));
    let sub_manager_addr = sub_account.manager.address()?;

    let new_account = create_default_account(&deployment.account_factory)?;
    // proxy and manager of different accounts
    let new_governance = GovernanceDetails::SubAccount {
        manager: sub_manager_addr.to_string(),
        proxy: proxy_addr.to_string(),
    };
    let err = new_account
        .manager
        .propose_owner(new_governance)
        .unwrap_err();
    let err = err.root().to_string();
    assert!(err.contains("manager and proxy has different account ids"));
    take_storage_snapshot!(chain, "account_move_ownership_to_falsy_sub_account");
    Ok(())
}

#[test]
fn account_updated_to_subaccount() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;

    // Creating account1
    let account = create_default_account(&deployment.account_factory)?;
    let proxy1_addr = account.proxy.address()?;
    let manager1_addr = account.manager.address()?;

    // Creating account2
    let account = create_default_account(&deployment.account_factory)?;
    let manager2_addr = account.manager.address()?;

    // Setting account1 as pending owner of account2
    account
        .manager
        .propose_owner(GovernanceDetails::SubAccount {
            manager: manager1_addr.to_string(),
            proxy: proxy1_addr.to_string(),
        })?;
    account.manager.set_address(&manager1_addr);
    account.proxy.set_address(&proxy1_addr);

    // account1 accepting account2 as a sub-account
    let accept_msg =
        abstract_core::manager::ExecuteMsg::UpdateOwnership(cw_ownable::Action::AcceptOwnership);
    account.manager.exec_on_module(
        to_json_binary(&abstract_core::proxy::ExecuteMsg::ModuleAction {
            msgs: vec![wasm_execute(manager2_addr, &accept_msg, vec![])?.into()],
        })?,
        PROXY.to_owned(),
        &[],
    )?;

    // Check manager knows about his new sub-account
    let ids = account.manager.sub_account_ids(None, None)?;
    assert_eq!(ids.sub_accounts.len(), 1);
    Ok(())
}

#[test]
fn account_updated_to_subaccount_recursive() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;

    // Creating account1
    let account = create_default_account(&deployment.account_factory)?;
    let proxy1_addr = account.proxy.address()?;
    let manager1_addr = account.manager.address()?;

    // Creating account2
    let account = create_default_account(&deployment.account_factory)?;

    // Setting account1 as pending owner of account2
    account
        .manager
        .propose_owner(GovernanceDetails::SubAccount {
            manager: manager1_addr.to_string(),
            proxy: proxy1_addr.to_string(),
        })?;
    // accepting ownership by sender instead of the manager
    account
        .manager
        .update_ownership(cw_ownable::Action::AcceptOwnership)?;

    // Check manager knows about his new sub-account
    account.manager.set_address(&manager1_addr);
    let ids = account.manager.sub_account_ids(None, None)?;
    assert_eq!(ids.sub_accounts.len(), 1);
    Ok(())
}

#[test]
fn top_level_owner() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;

    let account = create_default_account(&deployment.account_factory)?;
    // Creating sub account
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;
    let response = account.manager.sub_account_ids(None, None)?;
    let sub_account = AbstractAccount::new(&deployment, AccountId::local(response.sub_accounts[0]));

    let top_level_owner = sub_account.manager.top_level_owner()?;
    assert_eq!(top_level_owner.address, sender);
    Ok(())
}

#[test]
fn cant_renounce_with_sub_accounts() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;

    let account = create_default_account(&deployment.account_factory)?;
    // Creating sub account
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;

    let err: ManagerError = account
        .manager
        .update_ownership(cw_ownable::Action::RenounceOwnership)
        .unwrap_err()
        .downcast()
        .unwrap();
    assert_eq!(err, ManagerError::RenounceWithSubAccount {});
    Ok(())
}

#[test]
fn can_renounce_sub_accounts() -> AResult {
    let sender = Addr::unchecked(OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;

    let account = create_default_account(&deployment.account_factory)?;
    // Creating sub account
    account.manager.create_sub_account(
        vec![],
        "My subaccount".to_string(),
        None,
        None,
        None,
        None,
        None,
        &[],
    )?;

    let sub_account_id = account.manager.sub_account_ids(None, None)?.sub_accounts[0];

    let sub_account = AbstractAccount::new(&deployment, AccountId::local(sub_account_id));

    sub_account
        .manager
        .update_ownership(cw_ownable::Action::RenounceOwnership)?;

    account
        .manager
        .update_ownership(cw_ownable::Action::RenounceOwnership)?;

    // No owners
    let account_owner = account.manager.ownership()?;
    assert!(account_owner.owner.is_none());
    let sub_account_owner = sub_account.manager.ownership()?;
    assert!(sub_account_owner.owner.is_none());

    // Renounced governance
    let account_info = account.manager.info()?;
    assert_eq!(
        account_info.info.governance_details,
        GovernanceDetails::Renounced {}
    );
    let sub_account_info = sub_account.manager.info()?;
    assert_eq!(
        sub_account_info.info.governance_details,
        GovernanceDetails::Renounced {}
    );
    Ok(())
}
