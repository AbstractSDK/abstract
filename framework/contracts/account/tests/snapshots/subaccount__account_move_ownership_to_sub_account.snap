---
source: contracts/account/manager/tests/subaccount.rs
expression: all_storage
---
"abstract:account-factory":
  - - acseq
    - "4"
  - - cfg
    - "{\"version_control_contract\":\"contract1\",\"ans_host_contract\":\"contract0\",\"module_factory_address\":\"contract2\",\"ibc_host\":\"contract5\"}"
  - - contract_info
    - "{\"contract\":\"abstract:account-factory\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ans-host":
  - - config
    - "{\"next_unique_pool_id\":1}"
  - - contract_info
    - "{\"contract\":\"abstract:ans-host\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
  - - registered_dexes
    - "[]"
"abstract:ibc-client":
  - - config
    - "{\"version_control\":{\"address\":\"contract1\"},\"ans_host\":{\"address\":\"contract0\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-client\",\"version\":\"0.19.2\"}"
  - - module_data
    - "{\"module\":\"abstract:ibc-client\",\"version\":\"0.19.2\",\"dependencies\":[],\"metadata\":null}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-host":
  - - cfg
    - "{\"ans_host\":{\"address\":\"contract0\"},\"account_factory\":\"contract3\",\"version_control\":{\"address\":\"contract1\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-host\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-0":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"v\\u0005 W$ÛÒyø<\\\\wÃ\\u0001\\u001Bµg)[3>)¼â\\u001A\\u0002&·ù\\\"m\""
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"Monarchy\":{\"monarch\":\"owner\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-1":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"ØJ\\\\Ì1¬´B´ÛWn\\\"\\u001B\\u0012`±¿9½\\u0005\\u0002Íºñù¬\""
  - - "\u0000\bsub_accs\u0000\u0000\u0000\u0002"
    - "{}"
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"Monarchy\":{\"monarch\":\"owner\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":1}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-2":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\", ù\\u0000¾w¢»{\\n¿O¶wÄü8c\\u0003CÚ¼\\u000F´KêE\\u001D\""
  - - "\u0000\bsub_accs\u0000\u0000\u0000\u0003"
    - "{}"
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"My subaccount\",\"governance_details\":{\"SubAccount\":{\"manager\":\"úÃº\\u000F¡f9Ê\\u000E/´K}\\u001D)¨\\u00167ÐvNÚ¡ëã}|d\",\"proxy\":\"ØJ\\\\Ì1¬´B´ÛWn\\\"\\u001B\\u0012`±¿9½\\u0005\\u0002Íºñù¬\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":2}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"ØJ\\\\Ì1¬´B´ÛWn\\\"\\u001B\\u0012`±¿9½\\u0005\\u0002Íºñù¬\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-3":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"N°v¸§Ý\\u0010EÔ\\u001B÷>iäc/#î£³|²iÿC\\u0013d[\\\"-x\""
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"SubAccount\":{\"manager\":\"\\u0010Õ&A?ÆÇðæ)áàpä:^cÒùhÚ±}³¼¶ê@K\\u0003\",\"proxy\":\", ù\\u0000¾w¢»{\\n¿O¶wÄü8c\\u0003CÚ¼\\u000F´KêE\\u001D\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":3}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\", ù\\u0000¾w¢»{\\n¿O¶wÄü8c\\u0003CÚ¼\\u000F´KêE\\u001D\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:module-factory":
  - - "\u0000{5}config"
    - "{\"version_control_address\":\"contract1\",\"ans_host_address\":\"contract0\"}"
  - - contract_info
    - "{\"contract\":\"abstract:module-factory\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:proxy-local-0":
  - - "\u0000{5}state"
    - "{\"modules\":[\"D?Ê\\u0014¬å[#^,¡4\\r\\u000E\\r¡éÓ¯:5Z\\u000EÑüF%¾á\\b\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - admin
    - "\"D?Ê\\u0014¬å[#^,¡4\\r\\u000E\\r¡éÓ¯:5Z\\u000EÑüF%¾á\\b\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.19.2\"}"
"abstract:proxy-local-1":
  - - "\u0000{5}state"
    - "{\"modules\":[\"úÃº\\u000F¡f9Ê\\u000E/´K}\\u001D)¨\\u00167ÐvNÚ¡ëã}|d\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":1}"
  - - admin
    - "\"úÃº\\u000F¡f9Ê\\u000E/´K}\\u001D)¨\\u00167ÐvNÚ¡ëã}|d\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.19.2\"}"
"abstract:proxy-local-2":
  - - "\u0000{5}state"
    - "{\"modules\":[\"\\u0010Õ&A?ÆÇðæ)áàpä:^cÒùhÚ±}³¼¶ê@K\\u0003\",\"mock_module\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":2}"
  - - admin
    - "\"\\u0010Õ&A?ÆÇðæ)áàpä:^cÒùhÚ±}³¼¶ê@K\\u0003\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.19.2\"}"
"abstract:proxy-local-3":
  - - "\u0000{5}state"
    - "{\"modules\":[\"÷%$\\\"»Ì%÷w-sÃü8¦O0\\u0002.\\u0003sfD2cqM\\u0010\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":3}"
  - - admin
    - "\"÷%$\\\"»Ì%÷w-sÃü8¦O0\\u0002.\\u0003sfD2cqM\\u0010\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.19.2\"}"
"abstract:version-control":
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0005proxy0.19.2"
    - "{\"account_base\":6}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0007manager0.19.2"
    - "{\"account_base\":5}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\bans-host0.19.2"
    - "{\"native\":\"contract0\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\nibc-client0.19.2"
    - "{\"native\":\"contract4\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000emodule-factory0.19.2"
    - "{\"native\":\"contract2\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000faccount-factory0.19.2"
    - "{\"native\":\"contract3\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000fversion-control0.19.2"
    - "{\"native\":\"contract1\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0000"
    - "{\"manager\":\"D?Ê\\u0014¬å[#^,¡4\\r\\u000E\\r¡éÓ¯:5Z\\u000EÑüF%¾á\\b\",\"proxy\":\"v\\u0005 W$ÛÒyø<\\\\wÃ\\u0001\\u001Bµg)[3>)¼â\\u001A\\u0002&·ù\\\"m\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0001"
    - "{\"manager\":\"úÃº\\u000F¡f9Ê\\u000E/´K}\\u001D)¨\\u00167ÐvNÚ¡ëã}|d\",\"proxy\":\"ØJ\\\\Ì1¬´B´ÛWn\\\"\\u001B\\u0012`±¿9½\\u0005\\u0002Íºñù¬\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0002"
    - "{\"manager\":\"\\u0010Õ&A?ÆÇðæ)áàpä:^cÒùhÚ±}³¼¶ê@K\\u0003\",\"proxy\":\", ù\\u0000¾w¢»{\\n¿O¶wÄü8c\\u0003CÚ¼\\u000F´KêE\\u001D\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0003"
    - "{\"manager\":\"÷%$\\\"»Ì%÷w-sÃü8¦O0\\u0002.\\u0003sfD2cqM\\u0010\",\"proxy\":\"N°v¸§Ý\\u0010EÔ\\u001B÷>iäc/#î£³|²iÿC\\u0013d[\\\"-x\"}"
  - - "\u0000\u0005nmspcabstract"
    - "{\"trace\":\"local\",\"seq\":0}"
  - - "\u0000\u0007nmspc_a\u0000\u0005local\u0000\u0004\u0000\u0000\u0000\u0000abstract"
    - "8"
  - - cfg
    - "{\"account_factory_address\":\"contract3\",\"allow_direct_module_registration_and_updates\":true,\"namespace_registration_fee\":null}"
  - - contract_info
    - "{\"contract\":\"abstract:version-control\",\"version\":\"0.19.2\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"

